Re-design import catalog management (#3103)

leamas opened the pull request:
Instead of the temporary catalog update done today when importing plugins save  imported metadata and make the "effective catalog"  to be the results of merging the downloaded catalog with the imported metadata. 

The imported metadata stays in place until the imported plugin is uninstalled. This resolves the problems related to that the used catalog is "out of sync" with the imported plugins. for example as of #3103

Since this sort of launches an "effective catalog" which adds new items to it also opens up possibilities to tackle #3223.

EDIT: The PR changes the behaviour if the "Import Plugin" button. An imported plugin is installed directly. While an imported plugin is installed it "mask" the catalog entry; the only way to install from the catalog is to first uninstall the imported plugin. While IMHO more intuitive this might confuse some users adapted to the current behaviour. 

EDIT2 In the process, remove some quite long & complicated duplicated code. Here is also some more general clean up in pluginmanager.cpp. Important changes makes the list of loaded plugins owned by the plugin_loader, making the scope for changes in this important data structure much smaller. New data types and plenty of `const` declarations makes the job. The doumented behaviour of getInstalled() which have been broken is restored

NOTE!! Using this PR requires an updated ocpn-plugins.xml, see [below](https://github.com/OpenCPN/OpenCPN/pull/3258#issuecomment-1572210296).

leamas commented:
As stated, still a WIP. Need to think about possible corner cases and also testing...

leamas commented:
As long as everything goes as planned, this should work as-is. However, if things go wrong I can see two possible situations.

The first is if we end up with an imported library file without matching metadata.  There is no way to tell the difference compared to any library without matching metadata. The current blacklisting scheme should thus handle it.

The second is if we end up with an imported metadata file without a matching library, for example if user removes the library manually. This is possible to check for and remove the offending metadata file. However, this requires quite large changes which I hesitate to do in this PR which already is big enough. The proper action here is probably to file a new issue once this is merged.

So, personally I think the corner cases are dealt with for now. Still needs more testing, though.

leamas commented:
Testing under way in https://github.com/OpenCPN/OpenCPN/issues/3103. Updating top entry

transmitterdan commented:
I tried this today on a "virgin" install.  Two of the in-built plugins (charddldr_pi & wmm_pi) load as normal.  However, the other two (grib_pi & dashboard_pi) do not appear in the list of plugins.  Their entries in the opencpn.ini file are missing.  I don't understand why two are found in opencpn.ini and the other two are not.

leamas commented:
Thanks for trying! 

I made a late patch today: 1eaddf86 . Was this present when you tested?

transmitterdan commented:
> Thanks for trying!
> 
> I made a late patch today: [1eaddf8](https://github.com/OpenCPN/OpenCPN/commit/1eaddf86b188cca3b38ddaafcbe924e1caf252ab) . Was this present when you tested?

Yes, I believe it has the most recent patch you added to the PR.

transmitterdan commented:
It seems the issue is that if the plugin is not found in the opencpn.ini config file then when the plugin loader uses the Get() function to retrieve the enabled boolean with default set to true the true value is not "stickey".  A later Get() returns false and then the plugin is unloaded.  A second bug seems to be that the plugin is never shown in the list of plugins so it can never be enabled by the user.  This is probably because there is no entry made in the opencpn.ini file for the plugin due to the failure to honor the default Get() request.

leamas commented:
Thanks for input!

Another piece in the puzzle is that the sort order is changed, I found my grib in the wrong place so to speak. More work, and I'm running out of time. But I'm on it.

leamas commented:
The configuration file is actually interesting:
```
[PlugIns/libgrib_pi.so]
bEnabled=0
```

That entry is plain wrong, should be `PlugIns/GRIB`, which would explain a lot. To be continued...

leamas commented:
Rebased to current master and with new changes. I have made smoke tests on native Linux, Flatpak and Windows.

There is now only two kind of plugins in the GUI: The list of installed plugins and the list of available in the catalog. 
For installed plugins, metadata is taken from the actual plugin using the API when available. Remaining metadata is taken from an  xml file which is matched to the metadata.

NOTE!! For uninstalled plugins, the only source of truth is the catalog. As a consequence, the catalog needs entries also for the system plugins chartdownloader, grib, wmm and dashboard. I'm attaching a version of ocpn-plugins.xml which can be used to test. The plan is to make a PR to the plugins project when this feels stable.

[ocpn-plugins.xml.gz](https://github.com/OpenCPN/OpenCPN/files/11626677/ocpn-plugins.xml.gz)

@Hakansv : If you have some time...

Hakansv commented:
@leamas 
I've rebuilt pr_3258 now.
I used again the radar_pi as an example.
1. Updated and installed from master catalog. 
The text "Imported" is there although false.
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/1297675f-4b2f-4c2f-afeb-e80e6c9a777d)
2. Imported 5.5.1. alpha. 
This is after the import and no further actions
PM and actual do not agree:
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/82d85057-dd00-45f1-87d9-5082574d1b62)
3. Tried to install the imported. Not found?
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/98e9322a-0190-45c2-8d47-c4211c8299cf)

Did I built the correct code?

leamas commented:
> Did I built the correct code?

The commit should be dedee92fe604

I have no tested exactly the same code since I have rebased dan's localWinBuild branch on top of my stuff. However, the only changes are to _buildwin_ and various CMakeLists.txt, so we should still have the same functionality.

For me, installing  and importing Radar works as expected. Sigh.

One possible explanation could be if you have some left over from previous tests. What happens if you uninstall Radar and all other imported plugins. And then  checks in _C: \ProgramData\opencpn\plugins\install_data\imports_. It should be empty, remove any file which might be there. 

After this you should have a sane state. What happens now if you install Radar from the catalog, imports it, etc?

leamas commented:
BTW:
  - You are using the ocpn-plugins.xml I attached, right?
  - Could in possibly help to rebuild from scratch if you havn't done that?

leamas commented:
The plot thickens... I have rebuilt dedee92 and everything looks fine: I can import radar_pi and install it from the catalog in all sorts of combinations, it looks all fine. My test setup as follows:

  - The opencpn package is available on [cloudsmith](https://dl.cloudsmith.io/public/alec-leamas/opencpn/raw/files/opencpn-msvc_5.7.1+2549.77fc323_setup.exe)
  -  ocpn-plugins.xml, see [above](https://github.com/OpenCPN/OpenCPN/pull/3258#issuecomment-1572210296)
  -  I'm using the radar plugin Radar-5.5.1-alpha+2304221749.0d91def_msvc-wx32-10.0.19044-win32.tar.gz; it's also available on [cloudsmith](https://dl.cloudsmith.io/public/alec-leamas/opencpn/raw/files/Radar-5.5.1-alpha+2304221749.0d91def_msvc-wx32-10.0.19044-win32.tar.gz)

I would really appreciate if someone besides me could test this. The one who writes the code is not in the best position to test.

Hakansv commented:
With your cloudsmith package I get the same result as my builds so no difference there.
When testing with you example of ocpn-plugins.xml the view of installed master version of radar_pi is correct.
After import of radar_pi 5.5.1-alpa the PM says 5.5.1-alpha (Imported) but actual is v5.5.0. See picture.
If I then click "Reinstall" I get the strange message shown in the same picture.

BTW: When I closed (your) O ocpn-plugins.xml was no longer present in the folder? Looking closer it has became a folder named "\ocpn-plugins.xml\" with the file ocpn-plugins.xml in it??

![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/24e76342-5936-4c38-9726-1b4f8b021a3c)

leamas commented:
I have not forgotten this, just running out of time...

leamas commented:
For the record: I  can reproduce this. 
1. uninstall radar
2. Install from catalog
3. Import plugin. Seems to work, but does not actually install the imported plugin.

rgleason commented:
Alec, I have noticed this but just move on and reinstall, and select the one I want. A lot of this while testing PI. Yes, could be a little more consistent.

leamas commented:
np, the ball is on my side

leamas commented:
Two new commits. The test case described [above](https://github.com/OpenCPN/OpenCPN/pull/3258#issuecomment-1597625879)  now works for me also on windows.

Further testing is as always appreciated. However, sailing these these days so there is long tunr-around times.

Hakansv commented:
Built and tested on a Win10 using VS2022.
Some observations:

1. First time entering to Plugins:
For radar_pi the previous install was from catalog master.
The PM reports v.5.5.0 (Imported). "Imported" is not correct. See pict. 1

2. Example Watchdog after "Update catalog master" and nothing else.
There's an available update and that's correct.
The installed version reported is not correct but instead it prints the available update version. See pict 2.

I think I stop here with these two examples. I'll test the import function later.
First of all I think the installed plugin version reported in the upper left field has to be correct. The two fields with plugin version have to show the same thing, right?  Pict 3

Pict.1
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/759a26b9-773e-4f87-ba1b-069ee39ab57b)

Pict. 2
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/22b96779-926f-4fca-826b-a77905b42b1b)

Pict. 3 
![bild](https://github.com/OpenCPN/OpenCPN/assets/7202854/3d5f7415-2d13-45b9-a51d-d0abdc1ba7c9)

leamas commented:
One or two days ashore; more work.

Rebased to current master. Testing:

  - Updated master catalog
  - Installed  Radar,  correctly reported as  _Radar v5.5.0_
  - Installed Watchdog, reported as _Watchdog  2.4.55.0_
  - Imported Radar, reported as  ![image](https://github.com/OpenCPN/OpenCPN/assets/405541/8901a51e-114b-4cba-82a2-c056fc3f4db4)
  - Uninstalled Radar, works as expected
  - Installed Radar from catalog, looks fine:  _Radar v5.5.0_

There are possible differences which could explain the differrent results:

  1. Code differences after updates and rebasing.
  2. Differences in build procedures (I make a complete NSIS package and tests that)
  3. Differences in the environment, in particular left-overs in the  _C:\ProgramData\openpcn\plugins\install_data\imports_ directory.
  

It would be helpful you could shed some light on the different results using this list. This would then include make a clean rebuild of current head 9f4b73d, uninstalling all imported plugins and remove all files in the directy _...\imports_, see above.

leamas commented:
@Hakansv : ^

leamas commented:
The watchdog mess is basically about differences between the XML metadata  and the runtime version. What's reported by the plugin manager is the XML metadata. That these donb't match is basically a Watchdog bug, most likely introduced by  the catalog update https://github.com/OpenCPN/plugins/commit/721af139c1548287a4b20dcf14b44c7209d66812

Hakansv commented:
I've made a clean build, based on current master, of #3258.
Here's my test log:
1. I removed the one file present in the directory ...\imports
2. Uninstalled plugins Watchdog and the imported radar_pi
3. Update master catalog and installed radar v5.5.0. -> Correct reported version!
4. Installed Watchdog, 2.4.55.0 from master catalog. -> Correct reported version!
5. Imported a local radar_pi. -> Correct reported as 5.5.2-alpha+........[Imported]
6. Imported a local ais-vd tarball. ->Correct reported as ais-vd-1.8.2+2304252051.8df34f4[Imported]
7. Uninstalled the imported radar_pi. Works fine! Installed radar_pi from master catalog. -> Correct version reported.
This now seems to work. Great job!

One issue:
Log flooding for not installed plugins:
```
20:03:48.297 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.299 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin AIS Radar view
20:03:48.305 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.305 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin AIS Radar view
20:03:48.325 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.328 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin AutoTrackRaymarine
20:03:48.331 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.331 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin AutoTrackRaymarine
20:03:48.356 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.359 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Autopilot Route
20:03:48.363 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.363 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Autopilot Route
20:03:48.384 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.386 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Autopilot Route
20:03:48.389 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.389 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Autopilot Route
20:03:48.411 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.416 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Calculator
20:03:48.420 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.420 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Calculator
20:03:48.443 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.445 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Celestial Navigation
20:03:48.448 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.448 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Celestial Navigation
20:03:48.474 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.477 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Climatology
20:03:48.481 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.481 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Climatology
20:03:48.504 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.508 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Climatology
20:03:48.512 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.512 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Climatology
20:03:48.533 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.535 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DR
20:03:48.539 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.539 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DR
20:03:48.563 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.566 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DashboardSK
20:03:48.569 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.570 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DashboardSK
20:03:48.591 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.593 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DashboardTactics
20:03:48.596 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.596 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin DashboardTactics
20:03:48.619 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.621 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin EarthExplorer
20:03:48.625 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.625 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin EarthExplorer
20:03:48.645 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.648 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Engine Dashboard
20:03:48.653 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.653 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Engine Dashboard
20:03:48.677 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.682 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin FindIt
20:03:48.686 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.686 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin FindIt
20:03:48.712 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.715 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin GPS Odometer
20:03:48.719 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.720 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin GPS Odometer
20:03:48.743 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.748 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin GPS Odometer
20:03:48.751 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.751 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin GPS Odometer
20:03:48.775 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.778 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin IACFleet
20:03:48.782 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.782 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin IACFleet
20:03:48.806 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.809 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Launcher
20:03:48.812 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.813 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Launcher
20:03:48.835 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.837 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Logbook
20:03:48.842 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.842 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Logbook
20:03:48.864 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.867 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NSK
20:03:48.870 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.870 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NSK
20:03:48.896 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.899 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NVCharts
20:03:48.902 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.902 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NVCharts
20:03:48.923 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.925 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NmeaConverter_pi
20:03:48.928 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.928 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin NmeaConverter_pi
20:03:48.949 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.953 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin OCPN Draw
20:03:48.957 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.958 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin OCPN Draw
20:03:48.980 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.982 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ObjSearch
20:03:48.985 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:48.985 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ObjSearch
20:03:49.007 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.010 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Plots
20:03:49.014 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.014 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Plots
20:03:49.033 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.036 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Projections
20:03:49.039 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.039 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Projections
20:03:49.060 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.064 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Raymarine autopilot
20:03:49.069 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.069 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Raymarine autopilot
20:03:49.091 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.093 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin S63
20:03:49.097 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.097 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin S63
20:03:49.126 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.128 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ShipDriver
20:03:49.131 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.132 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ShipDriver
20:03:49.156 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.160 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ShipDriver
20:03:49.164 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.164 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin ShipDriver
20:03:49.185 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.187 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin StatusBar
20:03:49.192 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.192 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin StatusBar
20:03:49.212 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.215 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Tactics
20:03:49.218 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.218 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Tactics
20:03:49.238 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.241 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Testplugin
20:03:49.245 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.245 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin Testplugin
20:03:49.266 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.268 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin UKTides
20:03:49.272 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.272 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin UKTides
20:03:49.292 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.295 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin javascript
20:03:49.299 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.299 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin javascript
20:03:49.319 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.322 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin otcurrent
20:03:49.325 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.325 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin otcurrent
20:03:49.345 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.349 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin otidalplan
20:03:49.352 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.353 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin otidalplan
20:03:49.373 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.375 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin pypilot
20:03:49.378 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.378 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin pypilot
20:03:49.397 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.400 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin sar
20:03:49.403 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.404 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin sar
20:03:49.424 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.426 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin squiddio
20:03:49.429 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.429 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin squiddio
20:03:49.449 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.452 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin vfkaps
20:03:49.455 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.456 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin vfkaps
20:03:49.477 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.479 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin vfkaps
20:03:49.482 MESSAGE plugin_loader.cpp:118 Attempt to get version for empty pic pointer
20:03:49.482 MESSAGE plugin_loader.cpp:313 Atttempt to update enabled non-existing plugin vfkaps
20:03:49.811 MESSAGE OCPNPlatform.cpp:1007 ChangeLocale: Language load for:  en_US


```

Hakansv commented:
BTW:
I all times get this ask for corrections from Git:
include/catalog_parser.h:77: trailing whitespace.
+  PluginMetadata()
src/catalog_parser.cpp:43: trailing whitespace.
+

leamas commented:
Many thanks for testing! We are almost there... Open issues:

1. More robust handling of imported metadata, make sure it's removed when installing from catalog. Should not be necessary in the normal case, but all cases are not normal.
2. Take a look and handle the messages flooding the log.
3. Take care of formatting issues reported by git.

leamas commented:
Rebased: edited old commit to fix 3. above, new commit for 1. and 2.

leamas commented:
No joy, found problems related  to the built-in "system" plugins.

leamas commented:
Added a new patch for system plugins.  All set.

bdbcat commented:
Ready to merge?

leamas commented:
Yes, if done now, I'm ashore for two or three  days to be able to handle possible problems.

transmitterdan commented:
@leamas I am seeing some issues on Windows.  It seems that if one of the 4 "core" plugins is "disabled" in the opencpn.ini file then there is no way for the user to enable it as it no longer appears in the list of plugins.

leamas commented:
@transmitterdan : I will try to reproduce tomorrow, please stay tuned.

transmitterdan commented:
> @transmitterdan : I will try to reproduce tomorrow, please stay tuned.

Thanks.  It's not just Windows.  Linux also has same issue.  Details:

1) Chart downloader does not show up in the Options->Charts tab list.  Should appear at the end.
2) Disable plugin WMM (or other core plugin). Then close/open O.  WMM is no longer shown in the list of plugins.
